### PR TITLE
EIP712 Update to support multiple transactions

### PR DESCRIPTION
**Pull request by DracoLi**
By default, ethermint’s EIP712 implementation does not allow for different message types to be include in a single transaction. In order to resolve this issue, we need to update the default EIP712 type schema that ethermint uses for transactions to a different format that allows different message types to be included in a single transaction.

This PR updates the type def schema for messages by using a whitelist of permitted msg types stored in evm params.

### CODE REVIEW

File: x/evm/types/params_test.go [hunk 1]
```diff
@@ -47,8 +47,25 @@ func TestParamsValidate(t *testing.T) {
 		},
 		{
 			"invalid chain config",
-			NewParams("ara", true, true, ChainConfig{}, 2929, 1884, 1344),
-			false,
+			NewParams("ara", true, true, ChainConfig{
+				HomesteadBlock: newIntPtr(-1),
+			}, 2929, 1884, 1344),
+			true,
```
Comment (DracoLi) [inline x/evm/types/params_test.go:53]: noticed the original test was wrong so updated it here to properly test its invalid case.

**Review by rhuairahrighairidh — CHANGES_REQUESTED** [review]
nice, the simplifications look good

File: app/ante/ante_test.go [hunk 2]
```diff
@@ -26,6 +26,28 @@ func (suite AnteTestSuite) TestAnteHandler() {
...
+		{
+			LegacyMsgType: "delegate",
+			ValueTypes: []evmtypes.EIP712MsgAttrType{
+				{Name: "delegator_address", Type: "string"},
+				{Name: "validator_address", Type: "string"},
+				{Name: "amount", Type: "Coin"},
+			},
+		},
+	}
+	suite.app.EvmKeeper.SetParams(suite.ctx, params)
```
[the remark this answers is not in the document]
Comment (DracoLi) [reply app/ante/ante_test.go:49]: we have one on kava that sends cdp, hard, and convert erc20 msgs. Are you referring to specifically sending a send and delegate too?

File: ethereum/eip712/eip712.go [hunk 3]
```diff
@@ -1,23 +1,50 @@
...
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/common/math"
 	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/signer/core/apitypes"
+
+	evmtypes "github.com/tharsis/ethermint/x/evm/types"
 )
 
+// EIP712SignBytes returns the bytes to sign for a transaction.
+func EIP712SignBytes(chainID string, accnum, sequence, timeout uint64, fee legacytx.StdFee, msgs []sdk.Msg, memo string) []byte {
```
[the remark this answers is not in the document]
Comment (DracoLi) [reply ethereum/eip712/eip712.go:19]: :+1:

File: ethereum/eip712/eip712.go [hunk 4]
```diff
@@ -94,356 +115,79 @@ type FeeDelegationOptions struct {
...
-			if field.Len() == 0 {
-				// skip empty collections from type mapping
-				continue
-			}
-
-			fieldType = fieldType.Elem()
-			field = field.Index(0)
-			isCollection = true
+		// get corresponding allowed msg from params
+		legacyMsgType := legacyMsg.Type()
```
[the remark this answers is not in the document]
Comment (DracoLi) [reply ethereum/eip712/eip712.go:135]: great idea. I'll update to use the proto name

**Review by rhuairahrighairidh — APPROVED** [review]
looks good

**Pull request merged**